File: gateways/openresty/lua/stackverse/proxy.lua
local http = require("resty.http")

local config = require("stackverse.config")
local logging = require("stackverse.logging")
local problem = require("stackverse.problem")
local telemetry = require("stackverse.telemetry")

local _M = {}

local hop_by_hop = {
  ["connection"] = true,
  ["keep-alive"] = true,
  ["proxy-authenticate"] = true,
  ["proxy-authorization"] = true,
  ["te"] = true,
  ["trailer"] = true,
  ["transfer-encoding"] = true,
  ["upgrade"] = true,
}

local stripped_request = {
  ["authorization"] = true,
  ["content-length"] = true,
  ["cookie"] = true,
  ["host"] = true,
  ["x-xsrf-token"] = true,
}

local stripped_response = {
  ["connection"] = true,
  ["keep-alive"] = true,
  ["proxy-authenticate"] = true,
  ["proxy-authorization"] = true,
  ["te"] = true,
  ["trailer"] = true,
  ["transfer-encoding"] = true,
  ["upgrade"] = true,
}

local function request_body()
  local method = ngx.req.get_method()
  if method == "GET" or method == "HEAD" then
    return nil
  end
  ngx.req.read_body()
  local body = ngx.req.get_body_data()
  if body then
    return body
  end
  local file = ngx.req.get_body_file()
  if not file then
    return nil
  end
  local handle = io.open(file, "rb")
  if not handle then
    return nil
  end
  local data = handle:read("*a")
  handle:close()
  return data
end

local function request_headers(access_token)
  local source = ngx.req.get_headers(0)
  local headers = {}
  for key, value in pairs(source) do
    local lower = tostring(key):lower()
    if not hop_by_hop[lower] and not stripped_request[lower] then
      headers[key] = value
    end
  end
  if access_token then
    headers["Authorization"] = "Bearer " .. access_token
  end
  local traceparent = telemetry.ensure_traceparent(headers)
  if traceparent then
    headers["traceparent"] = traceparent
  end
  return headers
end

local function copy_response_headers(headers)
  for key, value in pairs(headers or {}) do
    if not stripped_response[tostring(key):lower()] then
      ngx.header[key] = value
    end
  end
end

local function upstream_failure(api, dependency, status, detail)
  logging.event("error", "dependency_call_failed", "failure", dependency .. " upstream request failed", {
    dependency = dependency,
    error_code = dependency .. "_unavailable",
  })
  if api then
    return problem.write(status, status == 502 and "Bad Gateway" or "Service Unavailable", detail)
  end
  ngx.status = status
  ngx.header["Content-Type"] = "text/plain; charset=utf-8"
  ngx.print(detail)
  return ngx.exit(status)
end

function _M.request(base_url, dependency, access_token, api)
  local url = config.join_url(base_url, ngx.var.request_uri)
  local httpc = http.new()
  httpc:set_timeout(30000)

  local response, err = httpc:request_uri(url, {
    method = ngx.req.get_method(),
    body = request_body(),
    headers = request_headers(access_token),
    keepalive_timeout = 60000,
    ssl_verify = true,
  })

  if not response then
    return upstream_failure(api, dependency, 502, "The upstream service is unavailable.")
  end

  ngx.status = response.status
  copy_response_headers(response.headers)
  local method = ngx.req.get_method()
  if method == "HEAD" or response.status == 204 or response.status == 304 then
    return ngx.exit(response.status)
  end
  ngx.print(response.body or "")
  return ngx.exit(response.status)
end

return _M

File: gateways/rust/src/proxy.rs
use axum::body::Body;
use axum::http::{Request, Response, StatusCode};
use futures_util::TryStreamExt;
use opentelemetry::global;
use opentelemetry::propagation::Injector;
use tracing_opentelemetry::OpenTelemetrySpanExt;
use url::Url;

use crate::AppState;
use crate::problems;

pub async fn proxy(
    state: &AppState,
    request: Request<Body>,
    target: &Url,
    dependency: &'static str,
    api: bool,
    access_token: Option<&str>,
) -> Response<Body> {
    let method = request.method().clone();
    let uri = request.uri().clone();
    let upstream_url = upstream_url(target, uri.path(), uri.query());
    let (parts, body) = request.into_parts();

    let mut headers = reqwest::header::HeaderMap::new();
    for (name, value) in parts.headers.iter() {
        if skip_request_header(name.as_str(), api) {
            continue;
        }
        headers.insert(name, value.clone());
    }
    if api {
        if let Some(token) = access_token {
            if let Ok(value) = reqwest::header::HeaderValue::from_str(&format!("Bearer {token}")) {
                headers.insert(reqwest::header::AUTHORIZATION, value);
            }
        }
    }
    inject_trace_context(&mut headers);

    let body = reqwest::Body::wrap_stream(body.into_data_stream());
    let method =
        reqwest::Method::from_bytes(method.as_str().as_bytes()).unwrap_or(reqwest::Method::GET);
    let response = state
        .http
        .request(method, upstream_url)
        .headers(headers)
        .body(body)
        .send()
        .await;

    match response {
        Ok(response) => upstream_response(response),
        Err(err) => {
            tracing::error!(
                event = "dependency_call_failed",
                outcome = "failure",
                dependency,
                error_code = %format!("{dependency}_unavailable"),
                error = %err,
                "Gateway upstream call failed"
            );
            if api {
                problems::problem(
                    StatusCode::BAD_GATEWAY,
                    "Bad Gateway",
                    "The upstream API is temporarily unavailable.",
                )
            } else {
                Response::builder()
                    .status(StatusCode::BAD_GATEWAY)
                    .header("content-type", "text/plain; charset=utf-8")
                    .body(Body::from("Bad Gateway"))
                    .expect("valid response")
            }
        }
    }
}

fn upstream_url(target: &Url, path: &str, query: Option<&str>) -> Url {
    let mut url = target.clone();
    url.set_path(path);
    url.set_query(query);
    url
}

fn upstream_response(response: reqwest::Response) -> Response<Body> {
    let status =
        StatusCode::from_u16(response.status().as_u16()).unwrap_or(StatusCode::BAD_GATEWAY);
    let mut builder = Response::builder().status(status);
    for (name, value) in response.headers() {
        if skip_response_header(name.as_str()) {
            continue;
        }
        builder = builder.header(name, value);
    }
    let stream = response
        .bytes_stream()
        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err));
    builder
        .body(Body::from_stream(stream))
        .expect("valid upstream response")
}

fn skip_request_header(name: &str, api: bool) -> bool {
    let lower = name.to_ascii_lowercase();
    if hop_by_hop(&lower) || lower == "host" || lower == "content-length" {
        return true;
    }
    if api {
        lower == "cookie" || lower == "authorization" || lower == "x-xsrf-token"
    } else {
        lower == "cookie"
    }
}

fn skip_response_header(name: &str) -> bool {
    let lower = name.to_ascii_lowercase();
    hop_by_hop(&lower)
}

fn hop_by_hop(name: &str) -> bool {
    matches!(
        name,
        "connection"
            | "keep-alive"
            | "proxy-authenticate"
            | "proxy-authorization"
            | "te"
            | "trailer"
            | "transfer-encoding"
            | "upgrade"
    )
}

fn inject_trace_context(headers: &mut reqwest::header::HeaderMap) {
    let context = tracing::Span::current().context();
    global::get_text_map_propagator(|propagator| {
        propagator.inject_context(&context, &mut HeaderInjector(headers))
    });
}

struct HeaderInjector<'a>(&'a mut reqwest::header::HeaderMap);

impl Injector for HeaderInjector<'_> {
    fn set(&mut self, key: &str, value: String) {
        if let (Ok(name), Ok(value)) = (
            reqwest::header::HeaderName::from_bytes(key.as_bytes()),
            reqwest::header::HeaderValue::from_str(&value),
        ) {
            self.0.insert(name, value);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::upstream_url;

    #[test]
    fn upstream_url_preserves_path_and_query() {
        let target = "http://backend:8080".parse().unwrap();
        assert_eq!(
            upstream_url(&target, "/api/v2/bookmarks", Some("visibility=public")).as_str(),
            "http://backend:8080/api/v2/bookmarks?visibility=public"
        );
    }
}
